### PR TITLE
pageserver: fix cont lsn jump on vectored read path

### PR DESCRIPTION
**Pull request by VladLazar**
## Problem
Vectored read path may return an image that's newer than the request lsn under certain circumstances.
```
  LSN
    ^
    |
    |
500 | ------------------------- -> branch point
400 |        X
300 |        X
200 | ------------------------------------> requested lsn
100 |        X
    |---------------------------------> Key

Legend:
* X - page images
```

The vectored read path inspects each ancestor timeline one by one starting from the current one.
When moving into the ancestor timeline, the current code resets the current search lsn (called `cont_lsn` in code)
to the lsn of the ancestor timeline ([here](https://github.com/neondatabase/neon/blob/d5708e74357ca19146098770895356326542306e/pageserver/src/tenant/timeline.rs#L2971)).

For instance, if the request lsn was 200, we would:
1. Look into the current timeline and find nothing for the key
2. Descend into the ancestor timeline and set `cont_lsn=500`
3. Return the page image at LSN 400

Myself and Christian find it very unlikely for this to have happened in prod since the vectored read path
is always used at the last record lsn.

This issue was found by a regress test during the work to migrate get page handling to use the vectored
implementation. I've applied my fix to that wip branch and it fixed the issue. 

## Summary of changes
The fix is to set the current search lsn to the min between the requested LSN and the ancestor lsn.
Hence, at step 2 above we would set the current search lsn to 200 and ignore the images above that.

A test illustrating the bug is also included. Fails without the patch and passes with it.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -4794,6 +4796,174 @@ mod tests {
...
+    //
+    // Legend:
+    // * X - page images
+    // ```
+    #[tokio::test]
+    async fn test_get_vectored_ancestor_descent() -> anyhow::Result<()> {
+        let tenant_conf = TenantConf {
+            // Make compaction deterministic
+            gc_period: Duration::ZERO,
+            compaction_period: Duration::ZERO,
```
Comment (koivunej) [inline pageserver/src/tenant.rs:4830]: Unless something has been changed elsewhere, tenant harness does not fire up the background tasks, it only marks the tenant as active. Checking..

File: pageserver/src/tenant.rs [hunk 2]
```diff
@@ -4794,6 +4796,174 @@ mod tests {
...
+
+        let child_timeline_id = TimelineId::generate();
+
+        tenant
+            .branch_timeline_test(&parent_timeline, child_timeline_id, Some(current_lsn), &ctx)
+            .await?;
+
+        let child_timeline = tenant
+            .get_timeline(child_timeline_id, true)
+            .expect("Should have the branched timeline");
```
Comment (koivunej) [inline pageserver/src/tenant.rs:4887]:
```suggestion
        let child_timeline = tenant
            .branch_timeline_test(&parent_timeline, child_timeline_id, Some(current_lsn), &ctx)
            .await?;
```

Weirdly enough we have a lot of tests with the `create + get` pattern.

File: pageserver/src/tenant.rs [hunk 3]
```diff
@@ -4794,6 +4796,174 @@ mod tests {
...
+            key = key.next();
+        }
+
+        child_timeline.freeze_and_flush().await?;
+
+        let lsn_offsets: [i64; 5] = [-10, -1, 0, 1, 10];
+        let mut query_lsns = Vec::new();
+        for image_lsn in parent_gap_lsns.keys().rev() {
+            for offset in lsn_offsets {
+                query_lsns.push(Lsn(u64::try_from(i64::try_from(image_lsn.0)? + offset)?))
```
Comment (koivunej) [inline pageserver/src/tenant.rs:4918]: There is [`u64::checked_add_signed`](https://doc.rust-lang.org/stable/std/primitive.u64.html#method.checked_add_signed) as well.

File: pageserver/src/tenant.rs [hunk 4]
```diff
@@ -4794,6 +4796,174 @@ mod tests {
...
+
+                if key == child_gap_at_key {
+                    parent_gap_lsns.insert(current_lsn, image_value);
+                }
+
+                key = key.next();
+            }
+
+            parent_timeline.freeze_and_flush().await?;
+        }
```
Comment (koivunej) [inline pageserver/src/tenant.rs:4877]: Do the rest of these keys have a function?

If I understand correctly, we produce 3 image values for 1k keys here, but only ever read one of them. Child gets images for `1k - 1` keys.

File: pageserver/src/tenant.rs (hunk 1 above)
Comment (VladLazar) [reply pageserver/src/tenant.rs:4830]: I wrote the test by copying a similar one and reworking it, hence this artefact. Will remove.

**Comment by VladLazar** [comment]
> The test looks convincing enough, but I cannot make the connection to the actual vectored read path change (`min`).
> 
> Does the old Timeline::get_value_reconstruct_data have this problem? There, we move to ancestor if cont_lsn is less than ancestor_lsn, and cont_lsn is originally `request_lsn + 1`. Okay, I think these are now aligned.

It's quite tricky to see the alignment between the read paths since they're structured differently. I think going through the example in the cover letter is the best way of understanding the issue. The test produces a very similar setup, albeit with a lot of boilerplate. Let me know if you wish to chat about it - the bug is quite obvious once you know about it :)

File: pageserver/src/tenant.rs (hunk 3 above)
Comment (VladLazar) [reply pageserver/src/tenant.rs]: Thanks - didn't know about it

https://github.com/neondatabase/neon/pull/7412/commits/11a1d2334d49fca229fd7cf37da74fff02eb9634

File: pageserver/src/tenant.rs (hunk 2 above)
Comment (VladLazar) [reply pageserver/src/tenant.rs]: Good point

https://github.com/neondatabase/neon/pull/7412/commits/11a1d2334d49fca229fd7cf37da74fff02eb9634

File: pageserver/src/tenant.rs (hunk 1 above)
Comment (VladLazar) [reply pageserver/src/tenant.rs]: https://github.com/neondatabase/neon/pull/7412/commits/11a1d2334d49fca229fd7cf37da74fff02eb9634

**Review by jcsp — APPROVED** [review]

File: pageserver/src/tenant.rs (hunk 4 above)
Comment (VladLazar) [reply pageserver/src/tenant.rs:4870]: Merging to get some soak time before next week's release. Let me know if this was a strong request.

**Comment by VladLazar** [comment]
I'll merge now to get some validation for the change before next weeks release.

**Comment by hlinnaka** [comment]
Great explanation and test case!

It took me a while to understand why the non-vectored path is not suffering from the same bug. I believe it's because `get_reconstruct_data' has this in the loop:
```

            // Recurse into ancestor if needed
            if is_inherited_key(key) && Lsn(cont_lsn.0 - 1) <= timeline.ancestor_lsn {
                trace!(
                    "going into ancestor {}, cont_lsn is {}",
                    timeline.ancestor_lsn,
                    cont_lsn
                );

                timeline_owned = timeline.get_ready_ancestor_timeline(ctx).await?;
                timeline = &*timeline_owned;
                prev_lsn = None;
                continue 'outer;
            }
```
Wouldn't hurt to explicitly test for the same scenario in the non-vectored case too. Or maybe we already have a test for that?

It would be useful to cross-check that `get_reconstruct_data` produces the same result as `get_vectored`. For example, at the end of `get_reconstruct_data`, call `get_vectored` with a corresponding single-item KeySpace, and assert that the result is the same.

**Comment by koivunej** [comment]
We had some discussion with @problame about this change in slack thread: <https://neondb.slack.com/archives/C033RQ5SPDH/p1717665056195329> about `min(request_lsn + 1, ancestor_lsn + 1)` vs. `min(cont_lsn + 1, ancestor_lsn + 1)`.

**Pull request merged**